File: Platform/Platform.Web/src/shared/api/entityMetadata.ts
import { useQuery } from '@tanstack/react-query';
import { httpClient } from './httpClient';
import { ServicePath } from './servicePaths';
import { entityMetadataKeys } from './queryKeys';
import type { EntityReference } from '../types/EntityReference';

/**
 * Tüm entity'lerde ortak cross-cutting metadata (audit / owner / state).
 * Backend: Platform.Application.Common.Metadata.EntityMetadata ile birebir.
 * Entity'ye özel Detail DTO'sundan bağımsız; tek generic endpoint'ten okunur.
 */
export interface EntityMetadata {
  isActive: boolean;
  owner?: EntityReference | null;
  organization?: EntityReference | null;
  createdBy?: EntityReference | null;
  createdAt: string;
  updatedBy?: EntityReference | null;
  updatedAt?: string | null;
}

export const entityMetadataDataSource = {
  get: async (entityType: string, id: string): Promise<EntityMetadata> => {
    const response = await httpClient.post<EntityMetadata>(
      ServicePath.EntityMetadata.Get,
      { entityType, id },
    );
    return response.data;
  },
};

/**
 * Detail sayfası footer'ı için ortak metadata sorgusu. `new` modda (id yok)
 * çalışmaz. Audit/owner referans verisi nadiren değişir → uzun staleTime.
 */
export function useEntityMetadata(entityType: string | undefined, id: string | undefined) {
  return useQuery({
    queryKey: entityMetadataKeys.detail(entityType ?? 'none', id ?? 'none'),
    queryFn: () => entityMetadataDataSource.get(entityType!, id!),
    enabled: !!entityType && !!id,
    staleTime: 5 * 60 * 1000,
  });
}

File: Platform/Platform.Application/Common/Metadata/EntityMetadata.cs
using Platform.Application.Modals.Common;

namespace Platform.Application.Common.Metadata;

/// <summary>
/// Tüm entity'lerde ortak olan cross-cutting bilgiler (IBaseEntity / IOwnedEntity /
/// IAuditableEntity). Entity'ye özel Detail DTO'larından bağımsız; tek bir generic
/// endpoint (api/entity-metadata/get) üzerinden, entity tipinden bağımsız okunur.
/// Owner/Organization/CreatedBy/UpdatedBy EntityReference olarak çözülür (ham Guid değil).
/// </summary>
public sealed class EntityMetadata
{
    public bool IsActive { get; set; }

    public EntityReference? Owner { get; set; }
    public EntityReference? Organization { get; set; }

    public EntityReference? CreatedBy { get; set; }
    public DateTime CreatedAt { get; set; }
    public EntityReference? UpdatedBy { get; set; }
    public DateTime? UpdatedAt { get; set; }
}
